File: android/app/src/main/java/com/agentsanywhere/app/feature/sessiondetail/TimelineTimestamp.kt
package com.agentsanywhere.app.feature.sessiondetail

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Timestamp rendering for timeline rows.
 *
 * Every timeline item carries the moment the server first saw it. Showing that
 * lets a reader tell when a reply or an action began, which is otherwise
 * impossible to reconstruct from a transcript that can span hours.
 *
 * The value renders as `[MM-DD HH:mm]` in the device's own timezone, in the same
 * bracketed form the timeline uses for other metadata. Seconds are omitted
 * because they are never the interesting part of a transcript reading, and the
 * year is omitted so the stamp stays short next to each row.
 */
private val timelineTimestampFormatter: DateTimeFormatter =
    DateTimeFormatter.ofPattern("MM-dd HH:mm")

/**
 * Format an ISO-8601 instant as `[MM-DD HH:mm]`, or null when unusable.
 *
 * The fields are assembled from a formatter rather than a locale-aware style so
 * the numeric form stays stable regardless of device language: a reader
 * comparing two timestamps should not have to contend with a locale-specific
 * field order.
 */
internal fun formatTimelineTimestamp(
    value: String?,
    zone: ZoneId = ZoneId.systemDefault(),
): String? {
    if (value.isNullOrBlank()) return null
    val instant = runCatching { Instant.parse(value) }.getOrNull() ?: return null
    return "[${timelineTimestampFormatter.format(instant.atZone(zone))}]"
}

File: desktop-workbench/renderer/src/components/session/timeline-timestamp.ts
/**
 * Timestamp rendering for timeline entries.
 *
 * Every timeline item carries the moment the server first saw it. Showing that
 * lets a reader tell when a reply or an action began, which is otherwise
 * impossible to reconstruct from a transcript that can span hours.
 *
 * The value is rendered as `[YYYY-MM-DD HH:mm:ss]` in the reader's own timezone,
 * matching the bracketed form the timeline uses for other metadata.
 */

/** Format an ISO timestamp as `[YYYY-MM-DD HH:mm:ss]`, or null when unusable. */
export function formatTimelineTimestamp(value: string | null | undefined): string | null {
  if (!value) return null
  const parsed = new Date(value)
  if (Number.isNaN(parsed.getTime())) return null
  return `[${localDateTime(parsed)}]`
}

/**
 * Format a date in local time.
 *
 * Built from the individual fields rather than `toLocaleString`, because the
 * numeric form is stable across locales and the transcript is read as data
 * rather than prose — a reader comparing two timestamps should not have to
 * contend with a locale-specific ordering.
 */
export function localDateTime(date: Date): string {
  const pad = (value: number) => String(value).padStart(2, "0")
  return (
    `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}` +
    ` ${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`
  )
}
